apps/mac.c

```diff
@@ -0,0 +1,203 @@
+    {"out", OPT_OUT, '>', "Output to filename rather than stdout"},
+    {NULL}
+};
+
+static int mac_ctrl_string(EVP_MAC_CTX *ctx, const char *value)
+{
+    int rv;
+    char *stmp, *vtmp = NULL;
+    stmp = OPENSSL_strdup(value);
+    if (!stmp)
+        return -1;
+    vtmp = strchr(stmp, ':');
+    if (vtmp) {
+        *vtmp = 0;
+        vtmp++;
+    }
+    rv = EVP_MAC_ctrl_str(ctx, stmp, vtmp);
+    OPENSSL_free(stmp);
+    return rv;
+}
+
+int mac_main(int argc, char **argv)
+{
+    char *mac_name = NULL;
+    int ret = 1;
+    char *prog;
+    const EVP_MAC *mac = NULL;
+    OPTION_CHOICE o;
+    EVP_MAC_CTX *ctx = NULL;
+    STACK_OF(OPENSSL_STRING) *opts = NULL;
+    unsigned char *buf = NULL;
+    size_t len;
+    int i;
+    BIO *in = NULL, *out = NULL;
+    const char *outfile = NULL;
+    const char *infile = NULL;
+    int out_bin = 0;
+
+
+    prog = opt_init(argc, argv, mac_options);
+    buf = app_malloc(BUFSIZE, "I/O buffer");
+    while ((o = opt_next()) != OPT_EOF) {
+        switch (o) {
+        case OPT_EOF:
+        case OPT_ERR:
+opthelp:
+            BIO_printf(bio_err, "%s: Use -help for summary.\n", prog);
+            goto err;
+        case OPT_HELP:
+            opt_help(mac_options);
+            ret = 0;
+            goto err;
+        case OPT_BIN:
+            out_bin = 1;
+            break;
+        case OPT_OUT:
+            outfile = opt_arg();
+            break;
+        case OPT_MAC:
+            mac_name = opt_arg();
+            mac = EVP_get_macbyname(mac_name);
+            if (mac == NULL) {
+                BIO_printf(bio_err, "mac name '%s' not found\n", mac_name);
+                goto err;
+            }
+            break;
+        case OPT_MACOPT:
+            if (opts == NULL)
+                opts = sk_OPENSSL_STRING_new_null();
+            if (opts == NULL || !sk_OPENSSL_STRING_push(opts, opt_arg()))
+                goto opthelp;
+            break;
+        }
+    }
+    argc = opt_num_rest();
+    argv = opt_rest();
+
+    if (mac == NULL) {
+        BIO_printf(bio_err, "name must be specified\n");
+        goto opthelp;
+    }
+
+    ctx = EVP_MAC_CTX_new(mac);
+    if (ctx == NULL)
+        goto err;
+
+    if (opts != NULL) {
+        for (i = 0; i < sk_OPENSSL_STRING_num(opts); i++) {
+            char *opt = sk_OPENSSL_STRING_value(opts, i);
+            if (mac_ctrl_string(ctx, opt) <= 0) {
+                BIO_printf(bio_err, "MAC parameter error '%s'\n", opt);
+                ERR_print_errors(bio_err);
+                goto err;
+            }
+        }
+    }
+
+    out = bio_open_default(outfile, 'w', out_bin ? FORMAT_BINARY : FORMAT_TEXT);
+    if (out == NULL) {
+        BIO_printf(bio_err, "Failed to open outfile \n");
+        goto err;
+    }
+
+    if (!EVP_MAC_init(ctx)) {
+        BIO_printf(bio_err, "EVP_MAC_Init failed\n");
+        goto err;
+    }
+
+    if (argc == 0) {
+        infile = "stdin";
+        in = BIO_new(BIO_s_file());
```

Comment:
Test `in != NULL` ?

And the previous line of `infile = "stdin";` seems no actual effect.

Reply:
added check.. infile is just used for error output.